### PR TITLE
refactor: casl factory - origdatablocks

### PR DESCRIPTION
**Pull request by HayenNico**
## Description
Subsection of PR #2748 for origdatablocks. Will need some adjustment in case #2760 is merged first.

This unifies the origDatablocksEndpointAccess and origDatablocksInstanceAccess functions in CaslAbilityFactory, removes instance-level Action elements and adjusts the affected controllers to accommodate the change. The origdatablock-specific code is extracted into a separate module.

## Changes:

* Replace CaslAbilityFactory.origDatablocksInstanceAccess and CaslAbilityFactory.origDatablocksEndpointAccess with one function CaslAbilityFactory.origDatablockAccess
* Code for CaslAbilityFactory.origDatablockAccessis factored out into new module OrigDatablockAbility
* Remove all instance-level origdatablock Action elements, rename endpoint-level actions
* Adjust endpoint and instance auth logic in origdatablock v3 and v4 controllers
* Adjustments in datasets-access.service for origdatablock lookup (will need merge conflict resolution with datasets PR)

## Tests included

- [x] Included for each change/fix?
- [x] Passing? 

## Documentation
- [x] swagger documentation updated (required for API changes)
- [x] official documentation update

## Summary by Sourcery

Refactor origdatablock authorization to use a unified CASL ability and simplify permission checks across controllers and dataset access.

Enhancements:
- Introduce a dedicated OrigDatablockAbility service and wire it into CaslAbilityFactory as a single origDatablockAccess entry point.
- Consolidate origdatablock endpoint and instance permissions into higher-level actions and remove fine-grained origdatablock action variants from the global action enum.
- Update v3 and v4 origdatablock controllers to rely on the unified ability for create/read/update/delete checks and filter construction.
- Adjust datasets-access.service origdatablock lookup logic to use the new permission model and handle unauthenticated users explicitly.
- Extend the CASL module wiring to register the new OrigDatablockAbility provider.

### CODE REVIEW

**Comment by nitrosx** [comment]
# Code Review Report: `refactor-casl-factory-origdatablocks` Branch

---

## Overview

1. **Verify logic correctness**: The refactored authorization logic is logically sound. The new `OrigDatablockAbility` class properly implements cumulative permissions with clear hierarchical access control. Permission checks correctly cascade from unauthenticated -> authenticated -> special groups (CREATE_DATASET_GROUPS, ADMIN_GROUPS, DELETE_GROUPS).

2. **Check if all edge cases are handled**: Most edge cases are handled. The code accounts for unauthenticated users, various group memberships, and configuration options like `#all` for CREATE_DATASET_GROUPS. However, there are some edge cases around undefined/null accessGroups that were addressed in later commits.

3. **Summarize what the code touched by the changes does**: The changes refactor the origdatablocks authorization system by:
   - Creating a dedicated `OrigDatablockAbility` service class that centralizes all origdatablock permission logic
   - Simplifying the `Action` enum by removing 19 redundant origdatablock-specific actions
   - Updating controllers to use the new centralized authorization approach
   - Adding comprehensive documentation

4. **Assess whether the changes make sense**: The refactoring is well-justified. It reduces code duplication, improves maintainability, and follows the principle of single responsibility. The permission model is now more consistent and easier to understand.

5. **Identify any unreachable code**: No obvious unreachable code was found in the main refactoring. The changes appear to be clean and purposeful.

**Commits responsible**: `fe5c0a42` (main refactor), `bb20370f` (typing fixes), `c642cebb` (undefined accessGroups handling), `e2d66cde` (documentation)

---

## Code Changes

### Files Modified
| File | Change Type | Lines Changed |
|------|-------------|----------------|
| `docs/authorization/origdatablocks.md` | Added | +75 |
| `src/casl/abilities/origdatablocks.ability.ts` | Added | +122 |
| `src/casl/action.enum.ts` | Modified | -22 (removed redundant actions) |
| `src/casl/casl-ability.factory.ts` | Modified | -216 (removed origdatablock logic) |
| `src/casl/casl.module.ts` | Modified | +1 (added OrigDatablockAbility provider) |
| `src/datasets/datasets-access.service.ts` | Modified | ~0 (updated to use new system) |
| `src/origdatablocks/origdatablocks.controller.ts` | Modified | -142 (simplified authorization) |
| `src/origdatablocks/origdatablocks.v4.controller.ts` | Modified | -117 (simplified authorization) |

### Key Changes Analysis

**1. New OrigDatablockAbility Class (`src/casl/abilities/origdatablocks.ability.ts`)**
- **Assessment**: Necessary and well-structured. Centralizes all origdatablock authorization logic in one place, following the pattern used by other CASL abilities in the system.
- **Improvement Needed**: None significant. The class is well-designed with clear separation of concerns and good documentation.
- **Verdict**: Excellent implementation that improves code organization and maintainability.

**2. Action Enum Simplification (`src/casl/action.enum.ts`)**
- **Assessment**: Highly beneficial. Removed 19 redundant origdatablock-specific actions (`OrigdatablockCreateOwner`, `OrigdatablockReadManyPublic`, etc.) that were duplicated and replaced them with 4 core actions.
- **Improvement Needed**: Consider adding deprecation warnings if any external code might be using the old action names.
- **Verdict**: Good simplification that reduces complexity without losing functionality.

**3. CASL Factory Refactoring (`src/casl/casl-ability.factory.ts`)**
- **Assessment**: Necessary cleanup. Removed ~216 lines of origdatablock-specific authorization code that is now handled by the dedicated `OrigDatablockAbility` class.
- **Improvement Needed**: None. The refactoring is clean and follows the established pattern.
- **Verdict**: Well-executed extraction of responsibilities.

**4. Controller Updates**
- **Assessment**: Both `origdatablocks.controller.ts` and `origdatablocks.v4.controller.ts` were simplified by removing duplicate authorization logic and delegating to the centralized `OrigDatablockAbility`.
- **Improvement Needed**: Some controller methods could benefit from more consistent error handling patterns.
- **Verdict**: Good simplification that reduces code duplication.

**5. Documentation (`docs/authorization/origdatablocks.md`)**
- **Assessment**: Excellent addition. Provides comprehensive documentation of the permission model, permission matrix, and implementation notes.
- **Improvement Needed**: Consider adding examples of configuration and usage patterns.
- **Verdict**: Very good documentation that will help developers understand and maintain the system.

---

## Security Review

1. **Are there any potential injection vulnerabilities?**
   - **No**. The code uses parameterized queries through Mongoose and doesn't construct any raw database queries with user input. All user input is properly validated.

2. **Does this code expose any sensitive user data?**
   - **No**. The authorization system properly restricts access based on group membership and publication status. Sensitive operations (create, update, delete) require appropriate permissions.

3. **Are there instances of insecure API usage?**
   - **No**. The API endpoints properly use the `@UseGuards(PoliciesGuard)` and `@CheckPolicies` decorators to enforce authorization before any operations are performed.

4. **Could this code lead to an authentication bypass?**
   - **No**. The refactoring maintains the same security posture as before. Authentication is still required for all write operations, and read operations are properly restricted based on publication status and group membership.

**Commits responsible**: All security aspects were maintained across all commits.

---

## Test Coverage

### Current Test Coverage Status
- **Factory Tests**: `src/casl/casl-ability.factory.spec.ts` exists but has minimal coverage for the new origdatablock functionality (only tests DatasetLifecycleUpdate)
- **OrigDatablockAbility Tests**: No dedicated test file exists for the new `OrigDatablockAbility` class
- **Controller Tests**: Existing controller tests (`origdatablocks.controller.spec.ts`, `origdatablocks.v4.controller.spec.ts`) may need updates to work with the new authorization system

### Assessment
The test coverage for the new authorization system is **inadequate**. While the existing tests may pass, they don't specifically test:
1. The new `OrigDatablockAbility` class functionality
2. The permission matrix for different user types
3. Edge cases in the authorization logic
4. Integration between the new ability class and the controllers

### Suggestions for Improvement

1. **Create `origdatablocks.ability.spec.ts`** with tests for:
   - Unauthenticated user permissions
   - Authenticated user permissions
   - CREATE_DATASET_GROUPS member permissions
   - ADMIN_GROUPS member permissions
   - DELETE_GROUPS member permissions
   - Edge cases (null user, undefined accessGroups, empty groups)

2. **Update existing controller tests** to verify:
   - Proper integration with the new authorization system
   - Error handling for unauthorized access
   - Permission inheritance and cascading

3. **Add integration tests** for:
   - End-to-end authorization flows
   - Permission boundary conditions

**Relevant commits**: The lack of tests was introduced in `fe5c0a42` and not addressed in subsequent commits.

---

## Security Examples

This section remains empty as no security vulnerabilities were identified in the code review.

---

## Testing for Security Use Cases

Since no specific security vulnerabilities were identified, this section focuses on testing the authorization logic:

### Test Cases for Authorization Logic

1. **Unauthenticated Access Tests**
   - **File**: `src/casl/abilities/origdatablocks.ability.ts:39,40`
   - **Test**: Verify unauthenticated users can only read published origdatablocks
   - **Code**: Test `buildAbility(null)` returns abilities that only allow `OrigdatablockRead` on published items

2. **Authenticated User Access Tests**
   - **File**: `src/casl/abilities/origdatablocks.ability.ts:55-57`
   - **Test**: Verify authenticated users can read origdatablocks they own, have access to, or are published
   - **Code**: Test with user having specific group memberships

3. **CREATE_DATASET_GROUPS Tests**
   - **File**: `src/casl/abilities/origdatablocks.ability.ts:59-70`
   - **Test**: Verify users in CREATE_DATASET_GROUPS can create/update origdatablocks where they are the owner
   - **Code**: Test permission checks with configured group membership

4. **ADMIN_GROUPS Tests**
   - **File**: `src/casl/abilities/origdatablocks.ability.ts:97-106`
   - **Test**: Verify admin users have unrestricted access
   - **Code**: Test all actions are permitted for admin group members

5. **DELETE_GROUPS Tests**
   - **File**: `src/casl/abilities/origdatablocks.ability.ts:108-115`
   - **Test**: Verify delete group members can delete any origdatablock
   - **Code**: Test `OrigdatablockDelete` permission for delete group members

6. **Edge Case Tests**
   - **File**: `src/casl/abilities/origdatablocks.ability.ts:22-26,64,76`
   - **Test**: Handle undefined/null accessGroups configuration
   - **Code**: Test behavior when configuration returns undefined for accessGroups

---

## Summary

### Consolidated Findings

**Overall Assessment**: The refactoring is **well-executed and beneficial**. It successfully centralizes origdatablock authorization logic, reduces code duplication, simplifies the action enum, and adds comprehensive documentation.

**Key Strengths**:
- Clear separation of concerns with dedicated `OrigDatablockAbility` class
- Simplified and more maintainable action enum
- Consistent permission model that's easier to understand
- Comprehensive documentation of the authorization model
- Proper handling of edge cases (undefined accessGroups)
- Maintained security posture throughout the refactoring

**Areas for Improvement**:
- **Test Coverage**: Inadequate testing of the new authorization system. No dedicated tests for `OrigDatablockAbility` and limited coverage of the permission matrix.
- **Error Handling**: Some inconsistencies in error handling patterns between controllers could be standardized.

**Security**: No security vulnerabilities were identified. The refactoring maintains the same security posture as the original code.

**Recommendations**:
1. **High Priority**: Add comprehensive unit tests for `OrigDatablockAbility` class
2. **Medium Priority**: Update existing controller tests to verify new authorization integration
3. **Low Priority**: Consider adding deprecation warnings for removed action enum values

**Verdict**: The changes are **approved with the recommendation to add test coverage**. The refactoring improves code quality and maintainability without introducing new issues.

**Generated using Mistral AI**

File: src/casl/abilities/origdatablocks.ability.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+      user.currentGroups.some((g) =>
+        this.accessGroups?.createDatasetPrivileged.includes(g),
+      )
+    ) {
+      /**
+       * User belonging to CREATE_DATASET_PRIVILEGED_GROUPS
+       */
+      can(Action.OrigdatablockCreate, OrigDatablock);
+      can(Action.OrigdatablockUpdate, OrigDatablock, ifOwner);
+    }
```
Comment (nitrosx) [inline src/casl/abilities/origdatablocks.ability.ts:95]: I am ok to leave it as it is in this PR, although we should try to combine condition as much as we can.

**Review by nitrosx — APPROVED** [review]
